packages/node/src/policy-gen/policy-gen-context.js

```diff
@@ -162,55 +173,102 @@ export class PolicyGeneratorContext {
+      const location = this.renames[descriptor.compartment]
+      if (!location) {
+        throw new TypeError(
+          `Compartment ${hrLabel(this.canonicalName)}: Rename map missing location for referenced compartment ${hrPath(descriptor.compartment)}`
+        )
+      }
+      return this.#readPowers.fileURLToPath(
+        new URL(descriptor.module, location)
       )
     }
-    return this.#readPowers.fileURLToPath(
-      new URL(descriptorOrLocation.module, location)
-    )
   }
 
   /**
-   * Attempts to find the filepath of a specifier
+   * Attempts to resolve the filepath of `specifier` using the current
+   * compartment descriptor's
+   * {@link CompartmentDescriptor.modules module descriptors}.
    *
    * Strips trailing slashes from specifiers, since Endo does not store them
    * with trailing slashes
    *
    * @param {string} specifier
-   * @returns {string | undefined}
+   * @returns {string | null} Filepath, name of builtin, or if `null`, then
+   *   either the module is missing or we just don't need an import map entry
+   *   for it.
    */
-  getFilepath(specifier) {
+  resolve(specifier) {
+    const [filepaths, log, compartment] = [
+      this.#filepaths,
+      this.#log,
+      this.compartment,
+    ]
     specifier = specifier.replace(/\/$/, '')
 
-    const moduleDescriptor = this.compartment.modules[specifier]
-    if (isCompleteModuleDescriptor(moduleDescriptor)) {
-      // FIXME: why doesn't this type narrow properly?
-      return this.toPath(
-        /** @type {Required<ModuleDescriptor>} */ (moduleDescriptor)
+    if (filepaths.has(specifier)) {
+      const cachedFilepath = /** @type {string | null} */ (
+        filepaths.get(specifier)
       )
+      if (this.isBuiltin(specifier)) {
+        /* c8 ignore next */
+        if (cachedFilepath !== specifier) {
+          log.warning(
+            `Compartment ${hrLabel(compartment.label)}: builtin specifier ${hrLabel(specifier)} previously ` +
+              (cachedFilepath === null
+                ? `considered missing; this is a bug`
+                : `resolved as package at ${hrPath(cachedFilepath)}; manual remediation may be necessary`)
+          )
+        }
+      }
+      return cachedFilepath
     }
 
-    if (this.#missingModules) {
-      const { label } = this.compartment
-      const missingModules = this.#missingModules.get(label) ?? new Set()
-      missingModules.add(specifier)
-      this.#missingModules.set(label, missingModules)
-    } else {
-      this.#log.warning(
-        `Missing module descriptor for specifier "${specifier}" in compartment "${this.compartment.label}"`
+    if (specifier.startsWith('./')) {
+      // relative path
+      filepaths.set(specifier, null)
+      return null
+    } else if (specifier.startsWith('../')) {
+      log.warning(
+        `Compartment ${hrLabel(compartment.label)}: attempting to resolve relative path ${hrPath(specifier)} beyond its boundary; manual remediation necessary`
```

Comment:
This seems wrong

```js
// index.js
require('./foo/a.js')
require('./foo/../../../oops.js') // beyond boundary
```
```js
// a.js
require('../b.js') // not beyond boundary
````

`specifier.startsWith('.')` might be fine as a single condition.

Reply:
for whatever reason I think I thought that this code path only gets followed when we didn't have a specifier pointing to a module in the compartment's own package

- [ ] verify and/or fix

Reply:
That's possible, but I don't think we need to differentiate between those and issue a warning. Not a concern of this part of code.

Reply:
fair enough